### ISSUE TITLE
url 0.5.10 and 1.5.1 beta regression

### ISSUE THREAD

**Issue by est31** [rust-lang/rust#43684]
[url](https://crates.io/crates/url) versions 0.5.10 and 1.5.1 regressed from stable to beta ([full log 1](https://cargobomb-reports.s3.amazonaws.com/stable-1.19-beta-1.20.0-beta.1/beta-2017-07-18/reg/url-0.5.10/log.txt), [full log 2](https://cargobomb-reports.s3.amazonaws.com/stable-1.19-beta-1.20.0-beta.1/beta-2017-07-18/reg/url-1.5.1/log.txt)):

```
Aug 04 02:42:21.790 INFO kablam! error[E0560]: struct `__test::test::TestDesc` has no field named `allow_fail`
Aug 04 02:42:21.790 INFO kablam!   --> tests/unit.rs:20:1
Aug 04 02:42:21.790 INFO kablam!    |
Aug 04 02:42:21.790 INFO kablam! 20 | / fn size() {
Aug 04 02:42:21.790 INFO kablam! 21 | |     use std::mem::size_of;
Aug 04 02:42:21.790 INFO kablam! 22 | |     assert_eq!(size_of::<Url>(), size_of::<Option<Url>>());
Aug 04 02:42:21.790 INFO kablam! 23 | | }
Aug 04 02:42:21.790 INFO kablam!    | |_^ `__test::test::TestDesc` does not have this field
Aug 04 02:42:21.790 INFO kablam! 
```
cc @seanmonstar @SimonSapin @Hoverbear

**Comment by arielb1** [comment]
wtf? need to investigate

**Comment by SimonSapin** [comment]
**TL;DR**: this regression itself is not a bug in Rust. If something here is arguably a bug, it’s the loophole that allowed me to indirectly depend on unstable APIs on the stable channel.

This is https://github.com/servo/rust-url/issues/371. It was fixed in https://github.com/servo/rustc-test/pull/4 and https://github.com/servo/rust-url/pull/372, we haven’t published a new version on crates.io since. It only affects compiling *tests*, not the library itself.

https://github.com/servo/rustc-test 0.1 is a copy of an old version of Rust’s `src/libtest` with `#[unstable]` attributes removed and a some other small changes to make it run on stable. When declaring the former as Cargo dependency, it ends up being used instead of the latter for `extern crate test;`.

https://github.com/rust-lang/rust/pull/42219 added a new field to `test::TestDesc`, so rustc-test 0.2 conditionally compiles to include that field or not based on the rustc version and adds a constructor that works in both cases. This 0.2 change was needed because not all the relevant code is in the `test` crate. `src/libsyntax/test.rs` also generates a static `&[TestDescAndFn]` when `rustc --test` is used.

We’ve discussed all this before in https://internals.rust-lang.org/t/test-and-external-test-harnesses/3145. I now think that the future of “proper” external test harnesses involves procedural macros 2.0, but I’ve been waiting for them to be closer to stabilization to open that discussion.

**Comment by arielb1** [comment]
Is this also the cause of #43683 (in `unic-idna-mapping`)?

**Comment by SimonSapin** [comment]
Yes.

**Comment by arielb1** [comment]
The issue looks fixed in rust-url. Closing.
